porch/func/wrapper-server/main.go

```diff
@@ -0,0 +1,133 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"net"
+	"os"
+	"os/exec"
+
+	pb "github.com/GoogleContainerTools/kpt/porch/func/evaluator"
+	"github.com/spf13/cobra"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/health/grpc_health_v1"
+	"google.golang.org/grpc/status"
+	"k8s.io/klog/v2"
+)
+
+func main() {
+	op := &options{}
+	cmd := &cobra.Command{
+		Use:   "wrapper-server",
+		Short: "wrapper-server is a gRPC server that fronts a KRM function",
+		RunE: func(cmd *cobra.Command, args []string) error {
+			argsLenAtDash := cmd.ArgsLenAtDash()
+			if argsLenAtDash > -1 {
+				op.entrypoint = args[argsLenAtDash:]
+			}
+			return op.run()
+		},
+	}
+	cmd.Flags().IntVar(&op.port, "port", 9446, "The server port")
+	if err := cmd.Execute(); err != nil {
+		fmt.Fprintf(os.Stderr, "unexpected error: %v\n", err)
+		os.Exit(1)
+	}
+}
+
+type options struct {
+	port       int
+	entrypoint []string
+}
+
+func (o *options) run() error {
+	address := fmt.Sprintf(":%d", o.port)
+	lis, err := net.Listen("tcp", address)
+	if err != nil {
+		return fmt.Errorf("failed to listen: %w", err)
+	}
+
+	evaluator := &singleFunctionEvaluator{
+		entrypoint: o.entrypoint,
+	}
+
+	klog.Infof("Listening on %s", address)
+
+	// Start the gRPC server
+	server := grpc.NewServer()
+	pb.RegisterFunctionEvaluatorServer(server, evaluator)
+	healthService := NewHealthChecker()
+	grpc_health_v1.RegisterHealthServer(server, healthService)
```

Comment:
will this correctly determine a case when the main service is not serving `EvaluateFunction` RPCs? Are they independent or does gRPC library somehow figure out when the function evaluation server stops serving?

Reply:
Quoted from [GRPC Health Checking Protocol](https://github.com/grpc/grpc/blob/v1.15.0/doc/health-checking.md)

>Health checks are used to probe whether the server is able to handle rpcs.

I didn't find any asynchronous operations in `pb.RegisterFunctionEvaluatorServer`. We assume that if the health check service is up, the `EvaluateFunction` service is up too.
It seems it's a common practice in k8s: https://kubernetes.io/blog/2018/10/01/health-checking-grpc-servers-on-kubernetes/
Note that k8s native grpc health check is available in k8s 1.23+.

Reply:
thanks for looking into this.